### PR TITLE
SNOW-1617549: throw warning on python 3.8

### PR DESCRIPTION
**Pull request by sfc-gh-aling**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-1617549

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.
   - [x] I acknowledge that I have ensured my changes to be thread-safe. Follow the link for more information: [Thread-safe Developer Guidelines](https://docs.google.com/document/d/162d_i4zZ2AfcGRXojj0jByt8EUq-DrSHPPnTa4QvwbA/edit#bookmark=id.e82u4nekq80k)

3. Please describe how your code solves the related issue.

throw one time warning when importing snowpark library on python 3.8

effect:

on python 3.8, this warns deprecation message exactly once

```python
Python 3.8.16 (default, Mar  1 2023, 21:19:10) 
[Clang 14.0.6 ] :: Anaconda, Inc. on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from snowflake.snowpark import functions
<stdin>:1: DeprecationWarning: Snowpark Python Runtime Support for Python 3.8 will reach end-of-life on March 31, 2025. To ensure continued support, please upgrade your runtime environment to Python > 3.8 before this date. For more details, please refer to https://docs.snowflake.com/en/developer-guide/python-runtime-support-policy.
>>> from snowflake.snowpark import functions
>>> from snowflake.snowpark import functions
>>> exit()
```

on python 3.11, no deprecation warning

```python
Python 3.11.9 (main, Apr 19 2024, 11:44:45) [Clang 14.0.6 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from snowflake.snowpark import functions
>>> from snowflake.snowpark import functions
>>> from snowflake.snowpark import functions
>>> from snowflake.snowpark import functions
>>> exit()
```

### CODE REVIEW

File: src/snowflake/snowpark/__init__.py [hunk 1]
```diff
@@ -69,3 +72,22 @@
     WhenNotMatchedClause,
 )
 from snowflake.snowpark.window import Window, WindowSpec
+
+_deprecation_warning_msg = (
+    "Snowpark Python Runtime Support for Python 3.8 will reach end-of-life "
+    "on March 31, 2025. To ensure continued support, please upgrade "
+    "your runtime environment to Python > 3.8 before this date. "
+    "For more details, please refer "
+    "to https://docs.snowflake.com/en/developer-guide/python-runtime-support-policy."
```
Comment (sfc-gh-yixie) [inline src/snowflake/snowpark/__init__.py:81]: @sfc-gh-deliu , @sfc-gh-qding Please review this text.

Comment (sfc-gh-pspaeth) [reply src/snowflake/snowpark/__init__.py:81]: "Python > 3.8" seems kind of awkward. How about something like "upgrade your runtime environment to a newer version before this date."

Also, are there any versions greater than 3.8 that we don't yet fully support?

Comment (sfc-gh-aling) [reply src/snowflake/snowpark/__init__.py:81]: @sfc-gh-pspaeth thanks for the feedback!

> Also, are there any versions greater than 3.8 that we don't yet fully support?
we currently support 3.8, 3.9, 3.10, 3.11.
3.12 and 3.13 are not supported yet

Comment (sfc-gh-pspaeth) [reply src/snowflake/snowpark/__init__.py:81]: So we can't say greater than 3.8, we have to specify versions 3.9 - 3.11.

Comment (sfc-gh-aling) [reply src/snowflake/snowpark/__init__.py]: updated

**Review by sfc-gh-pspaeth — COMMENTED** [review]
We don't support all the latest versions of Python, so when we say "upgrade to 3.9 or later" we're suggesting we do. We need to state which versions we support.

**Comment by sfc-gh-aling** [comment]
@sfc-gh-pspaeth thanks!
I've flatten the python version part as per your comment, I will leave the rest untouched -- those words are copied from an internal doc to communicate the deprecation.

**Review by sfc-gh-deliu — APPROVED** [review]
The change and the warning message looks good to me. If I understand correctly, this warning is triggered whenever they run `import snowpark`?

**Comment by sfc-gh-aling** [comment]
@sfc-gh-deliu yes, to be precise, it's whenever snowpark module is imported, the following examples will be able to trigger the warnings:

```python
import snowflake.snowpark
from snowflake import snowpark
from snowflake.snowpark import functions
from snowflake.snowpark.functions import sum
```

however, the following examples won't because snowpark is not imported
```python
import snowflake
import snowflake.connector
```

**Review by sfc-gh-jdu — APPROVED** [review]
any test for it?

**Comment by sfc-gh-aling** [comment]
@sfc-gh-jdu updated

**Pull request merged**